### PR TITLE
Resurrect bulk download tool for UCSB's data science course

### PR DESCRIPTION
**Pull request by beets**
It's much simplified - uses /stat/set to get the latest values, and only for states and counties in the US. Also doesn't look the best...

image

### CODE REVIEW

File: server/templates/tools/timeline.html [hunk 1]
```diff
@@ -38,4 +38,5 @@
 
 {% block subfooter_extra %}
 <button id="download-link" class="btn-subfooter btn btn-dark" style="visibility: hidden">Download Data</button>
+<a id="bulk-download-link" href="javascript:void(0)" style="visibility: hidden">Bulk download</a>
```
Comment (shifucun) [inline server/templates/tools/timeline.html:41]: nit: Bulk Download

File: static/js/tools/timeline_bulk_download.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ */
+import axios from "axios";
+import { TimelineParams } from "./timeline_util";
+
+/**
+ * Downloads data for all places of a certain type for the specified ptpv's in
+ * the chart.
+ */
+function downloadBulkData(statVars, placeType) {
+  console.log(statVars, placeType);
```
Comment (shifucun) [inline static/js/tools/timeline_bulk_download.ts:24]: remove logging?

File: static/js/tools/timeline_bulk_download.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+ * limitations under the License.
+ */
+import axios from "axios";
+import { TimelineParams } from "./timeline_util";
+
+/**
+ * Downloads data for all places of a certain type for the specified ptpv's in
+ * the chart.
+ */
+function downloadBulkData(statVars, placeType) {
```
Comment (shifucun) [inline static/js/tools/timeline_bulk_download.ts:23]: add a param for ancestor place to make it more generic?

File: static/js/tools/timeline_bulk_download.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            // that falls out of the range of 2xx
+            alert(
+              "There was an error loading the statistics: " +
+                error.response.data.message
+            );
+            return;
+          }
+          if (error.request) {
+            // The request was made but no response was received
+            console.log(error.request);
```
Comment (shifucun) [inline static/js/tools/timeline_bulk_download.ts:58]: change "console.log" to "alert" here and below?

File: static/js/tools/timeline_bulk_download.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    for (const sv of statVars) {
+      row.push(data[sv]["stat"][place]["value"]);
+    }
+    results.push(row.join(","));
+  }
+  const csv = results.join("\n");
+  saveToFile("datacommons_data.csv", csv);
+}
+
+function saveToFile(filename, csv) {
```
Comment (shifucun) [inline static/js/tools/timeline_bulk_download.ts:87]: could this be a shared function with the regular download?

File: server/templates/tools/timeline.html (hunk 1 above)
Comment (beets) [reply server/templates/tools/timeline.html]: done

File: static/js/tools/timeline_bulk_download.ts (hunk 3 above)
Comment (beets) [reply static/js/tools/timeline_bulk_download.ts]: done

File: static/js/tools/timeline_bulk_download.ts (hunk 2 above)
Comment (beets) [reply static/js/tools/timeline_bulk_download.ts]: done

File: static/js/tools/timeline_bulk_download.ts (hunk 4 above)
Comment (beets) [reply static/js/tools/timeline_bulk_download.ts]: these aren't appropriate for alerts. but logging to console isn't helpful either (if we really wanted to see the errors, they are also in the network tab). removed.

File: static/js/tools/timeline_bulk_download.ts (hunk 5 above)
Comment (beets) [reply static/js/tools/timeline_bulk_download.ts]: thanks for the reminder - used the shared version in util.js

**Review by beets — COMMENTED** [review]
also added a spinner while downloading data. ptal

**Review by shifucun — APPROVED** [review]
thanks for the change!

**Pull request merged**